proposed/2023/package-source-mapping-actions-PMUI.md

```diff
@@ -0,0 +1,102 @@
+
+Once preview restore determines the newly installed package IDs (top-level and transitive), those IDs will be checked against existing mappings. If an existing mapping is found, no new mapping is written for that package.
+
+If a package is found in the Global Packages Folder (GPF), that package will be mapped to the selected source.
+
+If the customer consents to the preview result, the new package source mappings (top-level and transitive) will be written to the applicable `nuget.config`.
+
+**Example 1** 
+
+```xml
+<PackageReference Include="Serilog" Version="11.0.0"/>
+```
+
+```xml
+<packageSourceMapping>
+    <packagesource key="nuget.org">
+        <package pattern="Serilog" />
+        <package pattern="SerilogTransitiveA" />
+        <package pattern="SerilogTransitiveB" />
+        ...
+    </packageSource>
+</packageSourceMapping>
+```
+
+**Result:**
+
+In this example, the customer had `nuget.org` selected from the source dropdown. They selected the `Serilog` package on the installed tab, chose version `11.0.0`, and selected the checkbox to make a mapping from `Serilog` to `nuget.org`. A mapping was created for the top-level package and for two transitive packages, `SerilogTransitiveA` and `SerilogTransitiveB`, because the transitives did not have an existing mapping.
+
+## Drawbacks
+
+## Rationale and Alternatives
+
+On the previous spec, there was a pin icon next to the sources dropdown to pin a source instead of a checkbox in the details pane. Reviewing with the UX Board, the pin concept may be confusing since there was nothing to explain what it did and it's not near the action buttons. The pin icon is used for different actions already in VS, like to pin a file at the top of the screen. It would be confusing to have the same button do different tasks in different parts of VS.
+
+Here is the mockup for the pin icon:
+
+![Options 3](../../meta/resources/PackageSourceMapping/VS.png)
+
+## Prior Art
+
+## Unresolved Questions
+
+None
+
+## Future Possibilities 
+
+1. Currently, if an `Install`/`Update` cannot find transitive packages on the mapped source being mapped, the restore will fail. Reaching out to other configured sources and allowing the customer to select a different source for these packages may resolve this issue. However, care must be taken to avoid unintentionally leaking requested package IDs to these secondary sources when the customer hasn't agreed to such a query.
+
+1. If a package is found in the Global Packages Folder (GPF), the PM UI could look at the source in the `nupkg.metadata` file. A new mapping could either be created for this source auotmatically, or it could be presented to the customer in the Preview Window or another affordance.
```

Comment:
An idea would be:

* Use the mapping from the GPF if the source exists.
* Warning if the source doesn't. 

I think https://github.com/NuGet/Client.Engineering/issues/1166 getting implemented will complement this nicely.